chatkit-core/src/main/kotlin/com/pusher/chatkit/state/AuxiliaryState.kt

```diff
@@ -0,0 +1,24 @@
+package com.pusher.chatkit.state
+
+import com.pusher.chatkit.rooms.state.JoinedRoomsState
+
+internal data class LastChange<T>(val version: Int, val changedData: T)
+
+@Suppress("CopyWithoutNamedArguments") // nextVersion (legible) passes as first param many times
+internal data class AuxiliaryState(
+    val version: Int,
+    val joinedRoomsReceived: LastChange<JoinedRoomsState>?
```

Comment:
we need that per reducer, for consistency and simplicity (to not create more parallel types) it would be also better to use action type (renaming `changedData` to `action`, as action that has led to the change and that contains data associated with the change). then it would be best to have with fun in ChatkitStore accepting second param: auxiliaryState and AuxiliaryState having `with` fun for `Action` which would be a sealed hierarchy. It's a centralisation point in the code but at least the compiler will help us to not forget about tracking last change for any action and it's a simple central point which actually also has advantage of being able to see all possible actions in one place.

Reply:
Hmm, I don't think I understand what you're asking for here.

When an action is sent to the ChatkitState we want the `with` function to add the AuxiliaryState automatically so we don't have to write the same thing twice. I thought that currently happened with your refactoring work?

Reply:
I reckon a little code draft will be helpful. I have consulted it with @grzesiekko and he has agreed it makes sense when actions are modelled as internal types and there's single action per reducer.
```kotlin
internal data class LastChange<T>(val version: Int, val action: Action)

@Suppress("CopyWithoutNamedArguments") // nextVersion (legible) passes as first param many times
internal data class AuxiliaryState(
    val version: Int,
    val joinedRoomsReceived: LastChange<JoinedRoomsReceived>?,
    val joinedRoom: LastChange<JoinedRoom>?
) {

    companion object {
        fun initial() = AuxiliaryState(
            0,
            null,
            null
        )
    }

    fun with(action: Action) =
        when (action) {
            is JoinedRoomsReceived ->
                copy(nextVersion, joinedRoomsReceived = LastChange(nextVersion, action))
            is JoinedRoom ->
                copy(nextVersion, joinedRoom = LastChange(nextVersion, action))
        }

    private val nextVersion = version + 1
}
```

Reply:
👍  have implemented your patch which contains this change and used where needed.